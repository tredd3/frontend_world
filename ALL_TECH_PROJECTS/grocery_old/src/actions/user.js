import {
  SET_USER,
  SET_NEW_AUTH_TOKEN
} from "./constants";
import API_ROUTES from '../helper/network/api-routes';
import {
  fetchApi
} from '../helper/network/fetch';
import store from '../store';
let NewaccessToken;

export const setUser = (data) => {
  return {
    type: SET_USER,
    data: data
  }
}

export const setNewToken = (data) => {
  return {
    type: SET_NEW_AUTH_TOKEN,
    data: data
  }
}

export const addUser = data => {
  return dispatch => {
    return fetchApi({
      url: API_ROUTES.addUser,
      body: data
    }).then(response => {
      dispatch(setUser({ userId: response.Data.userId }));
    }).catch(e => {

    })
    // dispatch(setUser({
    //   userId: "9953"
    // }));
  }
}



export const updateUser = data => {
  return (dispatch, getState) => {
    let nameArr = data.name.split(" ");
    let firstName = nameArr[0];
    const reducer = (accumulator, currentValue, i) => { return (i === 0) ? '' : accumulator + ' ' + currentValue };
    let lastName = nameArr.length !== 1
      ? nameArr.length !== 2
        ? nameArr.reduce(reducer, null)
        : nameArr[1]
      : ""
    fetchApi({
      url: API_ROUTES.updateUser,
      body: {
        userId: getState().user.userId,
        emailId: data.email,
        dob: data.dob,
        gender: data.gender,
        firstName,
        lastName,
        mobileNumber: data.phoneNumber
      }
    })
      .then(response => {
        let user = {
          ...getState().user,
          ...data
        }
        localStorage.setItem('user', JSON.stringify(user));
        return dispatch(setUser({
          ...user
        }));
      })
      .catch(e => {
        console.log("Error", e)
      })
  }
}

export const getUser = data => {
  return dispatch => {
    fetchApi({ url: API_ROUTES.getUserDetail, body: {} })
      .then(response => {
        let webApp = process.env.APP_TYPE == "web" ? true : false;
        let user = {
          customerName: response.Data.customerName,
          firstName: response.Data.firstName,
          lastName: response.Data.lastName,
          userId: response.Data.userId,
          phoneNumber: response.Data.mobileNumber,
          login: true,
          gender: response.Data.gender,
          dob: response.Data.dob,
          email: response.Data.emailId,
        }
        if (webApp) {
          user.access_token = data.access_token;
          user.refresh_token = data.refresh_token;
          user.uuid = data.uuid;
          user.expiresAt = data.expiresAt;
        } else {
          user = { ...store.getState().user, ...user };
        }
        localStorage.setItem('user', JSON.stringify(user));
        return dispatch(setUser({
          ...user
        }));
      })
      .catch(e => {
        let MessageCode = e.Control && e.Control.MessageCode
        if (MessageCode && MessageCode === 404) {
          if (data.history) {
            data.history.push("/account/address/add/addUder");
          }
          // fetchApi({
          //   url: API_ROUTES.addUser,
          //   body: { mobileNumber: "7021143366", firstName: "MyJio", lastName: "User" }
          // }).then(response => {
          //     dispatch(setUser({ userId: response.Data.userId }));
          //   })
          //   .catch(e => {

          //   }) 
        }
      })
  }
}

const addUserToBackend = data => {

}

export const getNewAccessToken = () => {
  return (dispatch, getState) => {
    const AUTH_KEY = process.env.AUTHORISATION_VALUE;
    var params = {
      grant_type: "refresh_token",
      refresh_token: getState().user.refresh_token
    }
    var formBody = [];
    for (var property in params) {
      var encodedKey = encodeURIComponent(property);
      var encodedValue = encodeURIComponent(params[property]);
      formBody.push(encodedKey + "=" + encodedValue);
    }
    formBody = formBody.join("&");
    fetch(API_ROUTES.getNewAccessToken, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        "Authorization": `Basic ${AUTH_KEY}`,
      },
      body: formBody
    }).then(response => response.json())
      .then(response => {
        let userNewTokens = {
          access_token: response.access_token,
          refresh_token: response.refresh_token,
          expiresAt: Date.now() + (response.expires_in * 1000),
        }
        localStorage.setItem('user', JSON.stringify({ ...getState().user, ...userNewTokens }));
        return dispatch(setNewToken(userNewTokens));
      })
      .catch(e => {
        console.log("ERROR HAPPENED During intializing User ")
      })
  }
}
