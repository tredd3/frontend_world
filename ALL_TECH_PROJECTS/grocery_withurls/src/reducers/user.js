import {
  SET_USER, NEW_USER, SET_NEW_AUTH_TOKEN, ORDER_RATED
} from '../actions/constants';

let initialState = {
  // eslint-disable-next-line @typescript-eslint/camelcase
  access_token: null,
  // eslint-disable-next-line @typescript-eslint/camelcase
  refresh_token: null,
  firstName: '',
  lastName: '',
  userId: 0,
  phoneNumber: '',
  login: false,
  xApiKey: null,
  newUser: false,
  newUserErrorMessage: '',
  orderRated: false
};
const placeOrderDataObj = sessionStorage.getItem('placeOrderDataObj');
let sessionUser = null;
if (placeOrderDataObj) {
  sessionUser = JSON.parse(placeOrderDataObj).user;
}
initialState = sessionUser || initialState;

export default function user(state = initialState, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, ...action.data };
    case SET_NEW_AUTH_TOKEN:
      return { ...state, ...action.data };
    case NEW_USER:
      return { ...state, ...action.data };
    case ORDER_RATED:
      return { ...state, ...action.data };
    default:
      return state;
  }
}
