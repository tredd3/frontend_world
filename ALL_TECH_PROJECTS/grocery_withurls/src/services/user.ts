import promiseMemoize from 'promise-memoize';
import makeApiCall from './helpers/make-api-call';
import { apiRoutes } from './api-routes';
import { formatDate } from '../helpers/functional';
import { pivotAPIUser } from './helpers/pivots';
import { APIUser, UserPivot } from '../types';
import { genderStringToCode } from './helpers/generic';

export const getUser = promiseMemoize(async (): Promise<UserPivot> => {
  const user = await makeApiCall<{}, APIUser>(apiRoutes.getUserDetail);
  return pivotAPIUser(user);
});

type UpdateUserArgs = {
  userId?: number;
  firstName?: string;
  lastName?: string;
  email?: string;
  gender?: UserPivot['gender'];
  dob?: Date;
  lat?: number;
  lng?: number;
  pincode?: number;
  city?: string;
  storeId?: number;
  storeName?: string;
};

export const updateUser = async ({
  firstName, lastName, email, gender, dob, lat, lng, pincode, city, storeId, storeName
}: UpdateUserArgs) => {
  const userFromApi = await getUser();
  const { userId } = userFromApi;
  const isValidStoreId = storeId !== undefined;
  const isValidStoreName = storeName !== undefined;

  await makeApiCall(apiRoutes.updateUser, {
    body: {
      userId,
      ...(firstName && { firstName }),
      ...(lastName && { lastName }),
      ...(email ? { emailId: email } : { emailId: '' }),
      ...(gender && { gender: genderStringToCode(gender) }),
      ...(dob ? { dob: formatDate(dob) } : { dob: '' }),
      preferences: {
        ...userFromApi.preferences,
        ...(pincode && { pincode }),
        ...(lat && { latitude: lat }),
        ...(lng && { longitude: lng }),
        ...(isValidStoreId && { storeId }),
        ...(city && { city }),
        ...(isValidStoreName && { storeName })
      }
    }
  });

  getUser.clear();
};

type AddUserArgs = {
  firstName: string;
  lastName: string;
};

export const addUser = ({ firstName, lastName }: AddUserArgs) => (
  makeApiCall<AddUserArgs, void>(apiRoutes.addUser, {
    body: {
      firstName,
      lastName
    }
  })
);
