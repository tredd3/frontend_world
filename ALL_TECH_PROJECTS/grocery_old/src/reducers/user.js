import { SET_USER, SET_NEW_AUTH_TOKEN } from '../actions/constants';

let initialState = {
  access_token: null,
  refresh_token: null,
  firstName: "",
  lastName: "",
  userId: 0,
  phoneNumber: '',
  login: false,
  xApiKey: null
}
initialState = JSON.parse(localStorage.getItem('user')) || initialState;

export default function user(state = initialState, action) {
  switch (action.type) {
    case SET_USER:
      return { ...state, ...action.data }
    case SET_NEW_AUTH_TOKEN:
      return { ...state, ...action.data }
    default:
      return state;
  }
}
