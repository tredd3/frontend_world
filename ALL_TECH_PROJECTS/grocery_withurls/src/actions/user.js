import { SET_USER, NEW_USER, SET_NEW_AUTH_TOKEN } from './constants';
import { checkServiceability } from './address';
import { getUser as getUserService, addUser as addUserService } from '../services/user';
import { getConfig } from '../services/config';
import updateUserAction from './update-user';
import { APIError } from '../services/types';

export const updateUser = updateUserAction;

export const setUser = data => ({
  type: SET_USER,
  data
});

export const newUser = data => ({
  type: NEW_USER,
  data
});

export const setNewToken = data => ({
  type: SET_NEW_AUTH_TOKEN,
  data
});

export const setNewUser = () => newUser({ newUser: true });

export const addUser = ({ firstName, lastName }) => async dispatch => {
  try {
    await addUserService({ firstName, lastName });
    getUserService.clear();
    const user = await getUserService();
    dispatch(setUser(user));
    dispatch(newUser({ newUser: false, newUserErrorMessage: '' }));
    try {
      await getConfig();
    } catch (e) {
      console.log(e.message);
    }
  } catch (e) {
    if (e instanceof APIError) {
      dispatch(newUser({ newUserErrorMessage: e.message }));
    }
  }
};

export const getUser = () => async dispatch => {
  try {
    const user = await getUserService();
    return [
      dispatch(setUser(user)),
      dispatch(checkServiceability({ pincode: user.preferences.pincode }))
    ];
  } catch (e) {
    return [
      dispatch(setUser({}))
    ];
  }
};
