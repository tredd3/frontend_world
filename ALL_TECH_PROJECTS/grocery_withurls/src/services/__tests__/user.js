import { getUser, updateUser, addUser } from '../user';
import mockDB from '../../mocks/db.json';

jest.mock('../../intents', () => ({
  getTokens: () => ({
    apiKey: 'dummy-api-key',
    accessToken: 'dummy-access-token'
  })
}));

describe('user service calls', () => {
  beforeEach(fetch.resetMocks);
  afterEach(getUser.clear);

  it('getUser method should return the correct structure', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));

    const user = await getUser();
    expect(user).toMatchSnapshot();
  });

  it('updateUser method should return the correct structure', async () => {
    fetch.mockResponses([JSON.stringify(mockDB.user)], [JSON.stringify(mockDB.updateUser)]);

    const user = {
      firstName: 'firstName',
      lastName: 'lastName',
      email: 'email',
      gender: 'female',
      dob: new Date('2020-02-29'),
      lat: 'lat',
      lng: 'lng',
      pincode: 'pincode',
      city: 'city',
      storeId: 'storeId'
    };

    await updateUser(user);
    expect(JSON.parse(fetch.mock.calls[1][1].body).Request.Body).toMatchSnapshot();

    fetch.mockResponseOnce(JSON.stringify(mockDB.user));
    await getUser();
    expect(fetch.mock.calls.length).toBe(3);
  });

  it('addUser method should respond with correct structure', async () => {
    fetch.mockResponseOnce(JSON.stringify(mockDB.user));

    const user = await addUser({ firstName: 'firstName', lastName: 'lastName' });
    expect(user).toMatchSnapshot();
  });
});
