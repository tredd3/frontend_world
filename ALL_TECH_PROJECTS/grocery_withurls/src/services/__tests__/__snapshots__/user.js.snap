// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user service calls addUser method should respond with correct structure 1`] = `
Object {
  "addedDate": 1577696828229,
  "customerName": "Firstname Lastname",
  "deviceId": "ZFB2U1FFTVpPZUk6QVBBOTFiRm9HVDB2ampGVWdTQUJyaTRBX0EwcmRtQko5RTQzcW03UUVxNG9x
S0l5dzk3TnRRNWhjX3lWcm5SMzlXSUxhc0wxSGtpTkEzRk5TUWJzZDN2aENBcUhYWXY3a3RoYkZT
UGJXam1ZRWJBc0ZWNmh2ajZVVXVxM1oydGtSU3FQQ2tkNDlQUFQ=
",
  "deviceType": "ANDROID",
  "firstName": "Firstname",
  "lastName": "Lastname",
  "lastUpdatedDate": 1578137110334,
  "mobileNumber": 9100100100,
  "preferences": Object {
    "latitude": 19.11224091559444,
    "longitude": 19.11224091559444,
    "pincode": 401701,
  },
  "registrationDate": 1577696828229,
  "status": 1,
  "userId": 10005564,
}
`;

exports[`user service calls getUser method should return the correct structure 1`] = `
Object {
  "addedDate": 1577696828229,
  "customerName": "Firstname Lastname",
  "deviceId": "ZFB2U1FFTVpPZUk6QVBBOTFiRm9HVDB2ampGVWdTQUJyaTRBX0EwcmRtQko5RTQzcW03UUVxNG9x
S0l5dzk3TnRRNWhjX3lWcm5SMzlXSUxhc0wxSGtpTkEzRk5TUWJzZDN2aENBcUhYWXY3a3RoYkZT
UGJXam1ZRWJBc0ZWNmh2ajZVVXVxM1oydGtSU3FQQ2tkNDlQUFQ=
",
  "deviceType": "ANDROID",
  "email": undefined,
  "firstName": "Firstname",
  "gender": undefined,
  "lastName": "Lastname",
  "lastUpdatedDate": 1578137110334,
  "login": true,
  "phoneNumber": 9100100100,
  "preferences": Object {
    "lat": 19.11224091559444,
    "lng": 19.11224091559444,
    "pincode": 401701,
  },
  "registrationDate": 1577696828229,
  "status": 1,
  "userId": 10005564,
}
`;

exports[`user service calls updateUser method should return the correct structure 1`] = `
Object {
  "dob": "2020-02-29",
  "emailId": "email",
  "firstName": "firstName",
  "gender": 2,
  "lastName": "lastName",
  "preferences": Object {
    "city": "city",
    "lat": 19.11224091559444,
    "latitude": "lat",
    "lng": 19.11224091559444,
    "longitude": "lng",
    "pincode": "pincode",
    "storeId": "storeId",
  },
  "userId": 10005564,
}
`;
